context/app/api/client.py

```diff
@@ -173,6 +163,46 @@ def _get_mock_vitessce_conf(self):
         }
 
 
+def _get_entity_from_hits(hits, has_token=None, uuid=None, hbm_id=None):
+    '''
+    >>> _get_entity_from_hits(['fake-hit-1', 'fake-hit-2'])
+    Traceback (most recent call last):
+    ...
+    Exception: ID not unique; got 2 matches
+
+    >>> def error(f):
+    ...   try: f()
+    ...   except Exception as e: print(type(e).__name__)
+
+    >>> error(lambda: _get_entity_from_hits([], hbm_id='HBM123.XYZ.456'))
+    Forbidden
+
+    >>> error(lambda: _get_entity_from_hits([], uuid='0123456789abcdef0123456789abcdef'))
+    Forbidden
+
+    >>> error(lambda: _get_entity_from_hits([], uuid='0123456789abcdef0123456789abcdef',
+    ...       has_token=True))
+    NotFound
+
+    >>> error(lambda: _get_entity_from_hits([], uuid='too-short'))
+    NotFound
+
+    >>> _get_entity_from_hits([{'_source': 'fake-entity'}])
+    'fake-entity'
+
+    '''
+    if len(hits) == 0:
+        if (uuid and len(uuid) == 32 or hbm_id) and not has_token:
+            # Assume that the UUID is not yet published:
+            # UI will suggest logging in.
+            abort(403)
```

Comment:
Didn't we want to use a `404` in these cases?

Reply:
No: I think this comment comes from before we talked, but the idea is that most of the time, people were getting 404s because they just weren't logged in, and so (usually) returning a 403 and telling them to log in will point them in the right direction.

(We can't actually check whether a given ID exists or not in the unpublished state unless we're logged in. This is partly to prevent information leakage... and partially because it's simpler not to have to deal with we-know-this-exists-but-can't-say-anything-else.)